backend/pkg/commons/db2/raw/cache.go

```diff
@@ -0,0 +1,119 @@
+package raw
+
+import (
+	"encoding/json"
+	"sync"
+	"time"
+)
+
+const (
+	oneBlockTTL = 1 * time.Second
+	blocksTTL   = 30 * time.Second // default ttl, if read it will be deleted sooner
+)
+
+type MinimalBlock struct {
+	Result struct {
+		Hash string `json:"hash"`
+	} `json:"result"`
+}
+
+type CachedStore struct {
+	store StoreReader
+	// sync.Map with manual delete have better perf than freecache because we can handle this way a ttl < 1s
+	cache sync.Map
+
+	locks   map[string]*sync.RWMutex
+	mapLock sync.Mutex // to make the map safe concurrently
+}
+
+func WithCache(reader StoreReader) *CachedStore {
+	return &CachedStore{
+		store: reader,
+		locks: make(map[string]*sync.RWMutex),
+	}
+}
+
+func (c *CachedStore) lockBy(key string) func() {
+	c.mapLock.Lock()
+	defer c.mapLock.Unlock()
+
+	lock, found := c.locks[key]
+	if !found {
+		lock = &sync.RWMutex{}
+		c.locks[key] = lock
+		lock.Lock()
+		return lock.Unlock
+	}
+	lock.RLock()
+	return lock.RUnlock
+}
+
+func (c *CachedStore) ReadBlockByNumber(chainID uint64, number int64) (*FullBlockData, error) {
+	key := blockKey(chainID, number)
+
+	unlock := c.lockBy(key)
+	defer unlock()
+
+	v, ok := c.cache.Load(key)
+	if ok {
+		// once read ensure to delete it from the cache
+		go c.unCacheBlockAfter(key, "", oneBlockTTL)
+		return v.(*FullBlockData), nil
+	}
+	// TODO make warning not found in cache
```

Comment:
```suggestion
```

Reply:
No once we have a proper logger I want to log something here